File: netlogo-gui/src/main/app/common/FindDialog.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.app.common;

import javax.swing.JDialog;
import java.awt.Frame;
import java.awt.event.ActionEvent;

import org.nlogo.core.I18N;
import org.nlogo.swing.NonemptyTextFieldActionEnabler;
import org.nlogo.swing.NonemptyTextFieldButtonEnabler;
import org.nlogo.swing.UserAction;
import org.nlogo.swing.UserAction.KeyBindings$;

public strictfp class FindDialog
    extends JDialog
    implements java.awt.event.ActionListener {
  /// STATIC CODE

  private static FindDialog instance;

  public static void init(java.awt.Frame frame) {
    instance = new FindDialog(frame);
  }

  public static FindDialog getInstance() {
    if (instance == null) {
      throw new IllegalStateException("FindDialog was never initialized");
    }
    return instance;
  }

  public static void watch(javax.swing.text.JTextComponent target) {
    FIND_ACTION.setEnabled(true);
    FindDialog findInstance = getInstance();
    findInstance.target = target;
    findInstance.setReplaceEnabled(target.isEditable());
  }

  public static void dontWatch(javax.swing.text.JTextComponent target) {
    FindDialog findInstance = getInstance();
    findInstance.setVisible(false);
    FIND_ACTION.setEnabled(false);
  }

  /// ACTIONS

  public static final javax.swing.Action FIND_ACTION = new FindAction();
  public static final javax.swing.Action FIND_NEXT_ACTION = new FindNextAction();

  public static class FindAction
      extends javax.swing.text.TextAction {
    FindAction() {
      super(I18N.guiJ().get("menu.edit.find"));
      putValue
          (javax.swing.Action.SMALL_ICON,
              new javax.swing.ImageIcon
                  (FindDialog.class.getResource
                      ("/images/magnify.gif")));
      putValue(UserAction.ActionCategoryKey(), UserAction.EditCategory());
      putValue(UserAction.ActionGroupKey(),    UserAction.EditFindGroup());
      putValue(javax.swing.Action.ACCELERATOR_KEY,
          UserAction.KeyBindings$.MODULE$.keystroke('F', true, false));
      setEnabled(false);
    }

    public void actionPerformed(java.awt.event.ActionEvent e) {
      FindDialog.getInstance().setVisible(true);
      FindDialog.getInstance().findBox.requestFocus();
      FindDialog.getInstance().findBox.selectAll();
      // Setting find field by default to selected text
      FindDialog findDialog = getInstance();
      String selectedText = findDialog.target.getSelectedText();
      if(selectedText != null){
        FindDialog.getInstance().findBox.setText(selectedText);
      }
      FindDialog.getInstance().setLocation
          (instance.owner.getLocation().x + instance.owner.getWidth()
              - instance.getPreferredSize().width,
              instance.owner.getLocation().y + instance.owner.getHeight() / 2
                  - instance.getPreferredSize().height / 2);
      FindDialog.getInstance().notFoundLabel.setVisible(false);
    }
  }

  public static class FindNextAction
      extends javax.swing.text.TextAction {
    FindNextAction() {
      super(I18N.guiJ().get("menu.edit.findNext"));
      putValue
          (javax.swing.Action.SMALL_ICON,
              new javax.swing.ImageIcon
                  (FindDialog.class.getResource
                      ("/images/magnify.gif")));
      putValue(UserAction.ActionCategoryKey(), UserAction.EditCategory());
      putValue(UserAction.ActionGroupKey(),    UserAction.EditFindGroup());
      putValue(javax.swing.Action.ACCELERATOR_KEY,
          UserAction.KeyBindings$.MODULE$.keystroke('G', true, false));
      setEnabled(false);
    }

    public void actionPerformed(java.awt.event.ActionEvent e) {
      String search = getInstance().findBox.getText();
      if (!getInstance().next(search, getInstance().ignoreCaseCheckBox.isSelected(), getInstance().wrapAroundCheckBox.isSelected())) {
        java.awt.Toolkit.getDefaultToolkit().beep();
      }
    }
  }


  /// INSTANCE CODE

  //     the text component to search
  protected javax.swing.text.JTextComponent target;

  // gui compoenents
  private final javax.swing.JButton nextButton, prevButton;
  NonemptyTextFieldButtonEnabler nextEnabler, prevEnabler;
  private final javax.swing.JButton replaceButton, replaceAndFindButton, replaceAllButton;
  NonemptyTextFieldButtonEnabler replaceEnabler, replaceAndFindEnabler, replaceAllEnabler;
  private final javax.swing.JCheckBox ignoreCaseCheckBox;
  private final javax.swing.JCheckBox wrapAroundCheckBox;
  private final javax.swing.JTextField findBox;
  private final javax.swing.JTextField replaceBox;
  private final javax.swing.JLabel replaceLabel;
  private final javax.swing.JLabel notFoundLabel;
  private final Frame owner;

  private FindDialog(Frame owner) {
    super(owner, I18N.guiJ().get("dialog.find.title"), false);
    this.owner = owner;

    // initialize GUI elements

    // text fields
    org.nlogo.swing.TextFieldBox findPanel =
        new org.nlogo.swing.TextFieldBox(javax.swing.SwingConstants.LEFT);

    findBox = new org.nlogo.swing.TextField(25);
    findBox.setEditable(true);
    findPanel.addField(I18N.guiJ().get("dialog.find.find"), findBox);

    replaceBox = new org.nlogo.swing.TextField(25);
    replaceBox.setEditable(true);
    replaceLabel = new javax.swing.JLabel(I18N.guiJ().get("dialog.find.replaceWith"));
    findPanel.addField(replaceLabel, replaceBox);

    // options panel elements
    javax.swing.JPanel optionsPanel = new javax.swing.JPanel();
    optionsPanel.setLayout(new javax.swing.BoxLayout(optionsPanel, javax.swing.BoxLayout.X_AXIS));
    ignoreCaseCheckBox = new javax.swing.JCheckBox(I18N.guiJ().get("dialog.find.ignoreCase"), true);
    wrapAroundCheckBox = new javax.swing.JCheckBox(I18N.guiJ().get("dialog.find.wrapAround"), true);
    notFoundLabel = new javax.swing.JLabel(I18N.guiJ().get("dialog.find.notFound"));

    optionsPanel.add(ignoreCaseCheckBox);
    optionsPanel.add(javax.swing.Box.createHorizontalStrut(12));
    optionsPanel.add(wrapAroundCheckBox);
    optionsPanel.add(javax.swing.Box.createHorizontalStrut(24));
    optionsPanel.add(notFoundLabel);
    notFoundLabel.setVisible(false);

    // buttons
    nextButton = new javax.swing.JButton(I18N.guiJ().get("dialog.find.next"));
    prevButton = new javax.swing.JButton(I18N.guiJ().get("dialog.find.previous"));
    replaceAndFindButton = new javax.swing.JButton(I18N.guiJ().get("dialog.find.replaceAndFind"));
    replaceButton = new javax.swing.JButton(I18N.guiJ().get("dialog.find.replace"));
    replaceAllButton = new javax.swing.JButton(I18N.guiJ().get("dialog.find.replaceAll"));

    getRootPane().setDefaultButton(nextButton);

    nextButton.addActionListener(this);
    prevButton.addActionListener(this);
    replaceAndFindButton.addActionListener(this);
    replaceButton.addActionListener(this);
    replaceAllButton.addActionListener(this);

    nextEnabler = new NonemptyTextFieldButtonEnabler(nextButton);
    nextEnabler.addRequiredField(findBox);
    new NonemptyTextFieldActionEnabler(FIND_NEXT_ACTION).addRequiredField(findBox);
    prevEnabler = new NonemptyTextFieldButtonEnabler(prevButton);
    prevEnabler.addRequiredField(findBox);
    replaceEnabler = new NonemptyTextFieldButtonEnabler(replaceAndFindButton);
    replaceEnabler.addRequiredField(findBox);
    replaceAndFindEnabler = new NonemptyTextFieldButtonEnabler(replaceButton);
    replaceAndFindEnabler.addRequiredField(findBox);
    replaceAllEnabler = new NonemptyTextFieldButtonEnabler(replaceAllButton);
    replaceAllEnabler.addRequiredField(findBox);

    org.nlogo.swing.Utils.addEscKeyAction
        (this,
            new javax.swing.AbstractAction() {
              public void actionPerformed(java.awt.event.ActionEvent e) {
                setVisible(false);
              }
            });

    javax.swing.JPanel buttonPanel =
        new org.nlogo.swing.ButtonPanel
            (new javax.swing.JComponent[]
                {
                    nextButton,
                    prevButton,
                    replaceAndFindButton,
                    replaceButton,
                    replaceAllButton
                }
            );

    // borders and layout
    optionsPanel.setBorder(new javax.swing.border.EmptyBorder(8, 8, 8, 8));
    findPanel.setBorder(new javax.swing.border.EmptyBorder(16, 8, 8, 8));
    buttonPanel.setBorder(new javax.swing.border.EmptyBorder(16, 8, 8, 8));

    getContentPane().setLayout(new java.awt.BorderLayout());
    getContentPane().add(findPanel, java.awt.BorderLayout.NORTH);
    getContentPane().add(optionsPanel, java.awt.BorderLayout.CENTER);
    getContentPane().add(buttonPanel, java.awt.BorderLayout.SOUTH);

    pack();
    setResizable(false);
    setVisible(false);

  }

  public void actionPerformed(ActionEvent e) {
    String search = findBox.getText();

    if (e.getSource().equals(nextButton) || e.getSource().equals(findBox)) {
      if (!next(search, ignoreCaseCheckBox.isSelected(), wrapAroundCheckBox.isSelected())) {
        java.awt.Toolkit.getDefaultToolkit().beep();
        notFoundLabel.setVisible(true);
      } else {
        notFoundLabel.setVisible(false);
      }
    } else if (e.getSource().equals(prevButton)) {
      boolean ok =
          prev(search, ignoreCaseCheckBox.isSelected(), wrapAroundCheckBox.isSelected());
      if (!ok) {
        java.awt.Toolkit.getDefaultToolkit().beep();
        notFoundLabel.setVisible(true);
      } else {
        notFoundLabel.setVisible(false);
      }
    } else if (e.getSource().equals(replaceAndFindButton)) {
      // Stores if the Replace and Find is selected for the first time.
      // Also checks if the text in find field and selected text are same or not.
      boolean firstTime = target.getSelectedText() == null || (target.getSelectedText() != null
                && !(ignoreCaseCheckBox.isSelected() ? target.getSelectedText().equalsIgnoreCase(findBox.getText())
                : target.getSelectedText().equals(findBox.getText())));
      if (!firstTime) {
          replace(replaceBox.getText());
      }
      if (!next(search, ignoreCaseCheckBox.isSelected(), wrapAroundCheckBox.isSelected())) {
        java.awt.Toolkit.getDefaultToolkit().beep();
        notFoundLabel.setVisible(true);
      } else {
        notFoundLabel.setVisible(false);
      }
    } else if (e.getSource().equals(replaceButton)) {
      replace(replaceBox.getText());
    } else if (e.getSource().equals(replaceAllButton)) {
      replaceAll(search, ignoreCaseCheckBox.isSelected(), replaceBox.getText());
    } else {
      notFoundLabel.setVisible(false);
    }
  }

  private boolean next(String search, boolean ignoreCase, boolean wrapAround) {
    int start = target.getSelectionEnd();
    int matchIndex = -1;

    String text = getTargetText();

    if (ignoreCase) {
      // this might get slow with big programs. should be tested. -AZS
      search = search.toUpperCase();
      text = text.toUpperCase();
    }

    matchIndex = text.indexOf(search, start);

    if (matchIndex == -1 && wrapAround) {
      text = text.substring(0, start);
      matchIndex = text.indexOf(search);
    }

    if (matchIndex > -1) {
      target.setSelectionStart(matchIndex);
      target.setSelectionEnd(matchIndex + search.length());
      return true;
    } else {
      return false;
    }
  }

  private boolean prev(String search, boolean ignoreCase, boolean wrapAround) {
    int start = StrictMath.max(0, target.getSelectionStart() - 1);
    int matchIndex = -1;

    String text = getTargetText();

    if (ignoreCase) {
      // this might get slow with big programs. should be tested. -AZS
      search = search.toUpperCase();
      text = text.toUpperCase();
    }

    matchIndex = text.lastIndexOf(search, start);

    if (matchIndex == -1 && wrapAround) {
      text = text.substring(start, text.length());
      if (matchIndex != -1) {
        matchIndex = start + text.lastIndexOf(search);
      }
    }

    if (matchIndex > -1) {
      target.setSelectionStart(matchIndex);
      target.setSelectionEnd(matchIndex + search.length());
      return true;
    } else {
      return false;
    }
  }

  private void replace(String replacement) {
    if (target.getSelectedText() == null || target.getSelectedText().length() == 0) {
      java.awt.Toolkit.getDefaultToolkit().beep();
      return;
    }

    try {
      target.getDocument().remove(
          target.getSelectionStart(),
          target.getSelectionEnd() - target.getSelectionStart()
      );
      target.getDocument().insertString(target.getCaretPosition(), replacement, null);
    } catch (javax.swing.text.BadLocationException ex) {
      java.awt.Toolkit.getDefaultToolkit().beep();
    }
  }

  private int replaceAll(String search, boolean ignoreCase, String replacement) {
    target.setSelectionStart(0);
    target.setSelectionEnd(0);
    int j = 0;
    if (next(search, ignoreCase, false)) {
      j = 1;
      do {
        replace(replacement);
        j++;

        // arbitrary redundant check to avoid infinite loop
        if (j > 50000) {
          throw new IllegalStateException("Replace All replaced too many items.");
        }
      }
      // never wrap around on replace all
      while (next(search, ignoreCase, false));
    }
    return j;
  }

  private String getTargetText() {
    String text;

    if (target instanceof javax.swing.JEditorPane) {
      // we need to get the text this way to avoid returning the HTML
      // tags which screw-up the search - jrn 7/22/05
      try {
        text = target.getText(0, target.getDocument().getLength());
      } catch (javax.swing.text.BadLocationException ex) {
        throw new IllegalStateException(ex);
      }
    } else {
      text = target.getText();
    }
    return text;
  }

  // Disables/Enables all the replace buttons/textfield
  protected void setReplaceEnabled(boolean enabled) {
    replaceEnabler.setEnabled(enabled);
    replaceAndFindEnabler.setEnabled(enabled);
    replaceAllEnabler.setEnabled(enabled);
    replaceBox.setEnabled(enabled);
    replaceLabel.setEnabled(enabled);
  }

  static class FocusListener implements java.awt.event.FocusListener {
    @Override
    public void focusGained(java.awt.event.FocusEvent fe) {
      if (fe.getSource() instanceof javax.swing.text.JTextComponent) {
        FindDialog.watch((javax.swing.text.JTextComponent) fe.getSource());
      }
    }

    @Override
    public void focusLost(java.awt.event.FocusEvent fe) {
      if (! fe.isTemporary() && fe.getSource() instanceof javax.swing.text.JTextComponent) {
        FindDialog.dontWatch((javax.swing.text.JTextComponent) fe.getSource());
      }
    }
  }
}

File: netlogo-gui/src/main/app/common/FindDialog.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.app.common

import java.awt.{ BorderLayout, Frame, Toolkit }
import java.awt.event.{ ActionEvent, ActionListener, FocusEvent }
import java.util.Locale
import javax.swing.{ AbstractAction, Action, Box, BoxLayout, JDialog, JEditorPane, JLabel, JPanel, SwingConstants }
import javax.swing.border.EmptyBorder
import javax.swing.text.{ BadLocationException, TextAction }

import org.nlogo.core.I18N
import org.nlogo.theme.{ InterfaceColors, ThemeSync }
import org.nlogo.swing.{ ButtonPanel, CheckBox, DialogButton, NonemptyTextFieldActionEnabler,
                         NonemptyTextFieldButtonEnabler, ScrollableTextComponent, TextField, TextFieldBox, Transparent,
                         UserAction, Utils },
  UserAction.{ EditCategory, EditFindGroup, KeyBindings, MenuAction }

object FindDialog extends ThemeSync {
  class FindAction extends TextAction(I18N.gui.get("menu.edit.find")) with MenuAction {
    category = EditCategory
    group = EditFindGroup
    accelerator = KeyBindings.keystroke('F', withMenu = true)

    setEnabled(false)

    def actionPerformed(e: ActionEvent): Unit = {
      val codeInstance = getCodeInstance

      if (codeInstance.isVisible)
        codeInstance.setVisible(false)

      val instance = getInstance

      instance.setVisible(true)
      instance.findBox.requestFocus()
      instance.findBox.selectAll()

      instance.findBox.setText(instance.target.flatMap(t => Option(t.getSelectedText))
                                              .getOrElse(codeInstance.findBox.getText))

      instance.setLocation(instance.owner.getLocation.x + instance.owner.getWidth - instance.getPreferredSize.width,
                           instance.owner.getLocation.y + instance.owner.getHeight / 2 -
                           instance.getPreferredSize.height / 2)
      instance.notFoundLabel.setVisible(false)
    }
  }

  class FindNextAction extends TextAction(I18N.gui.get("menu.edit.findNext")) with MenuAction {
    category = EditCategory
    group = EditFindGroup
    accelerator = KeyBindings.keystroke('G', withMenu = true)

    setEnabled(false)

    def actionPerformed(e: ActionEvent): Unit = {
      if (!getInstance.next(getInstance.findBox.getText, getInstance.ignoreCaseCheckBox.isSelected,
                            getInstance.wrapAroundCheckBox.isSelected))
        Toolkit.getDefaultToolkit.beep()
    }
  }

  class FindActionCode extends TextAction(I18N.gui.get("menu.edit.find")) with MenuAction {
    category = EditCategory
    group = EditFindGroup

    setEnabled(false)

    def actionPerformed(e: ActionEvent): Unit = {
      val instance = getInstance

      if (instance.isVisible)
        instance.setVisible(false)

      val codeInstance = getCodeInstance

      codeInstance.setVisible(true)
      codeInstance.findBox.requestFocus()
      codeInstance.findBox.selectAll()

      codeInstance.findBox.setText(codeInstance.target.flatMap(t => Option(t.getSelectedText))
                                              .getOrElse(instance.findBox.getText))

      codeInstance.setLocation(codeInstance.owner.getLocation.x + codeInstance.owner.getWidth -
                               codeInstance.getPreferredSize.width, codeInstance.owner.getLocation.y +
                               codeInstance.owner.getHeight / 2 - codeInstance.getPreferredSize.height / 2)
      codeInstance.notFoundLabel.setVisible(false)
    }
  }

  class FindNextActionCode extends TextAction(I18N.gui.get("menu.edit.findNext")) with MenuAction {
    category = EditCategory
    group = EditFindGroup
    accelerator = KeyBindings.keystroke('G', withMenu = true)

    setEnabled(false)

    def actionPerformed(e: ActionEvent): Unit = {
      if (!getCodeInstance.next(getCodeInstance.findBox.getText, getCodeInstance.ignoreCaseCheckBox.isSelected,
                                getCodeInstance.wrapAroundCheckBox.isSelected))
        Toolkit.getDefaultToolkit.beep()
    }
  }

  class FocusListener extends java.awt.event.FocusListener {
    def focusGained(e: FocusEvent): Unit = {
      e.getSource match {
        case s: ScrollableTextComponent =>
          watch(s)

        case _ =>
      }
    }

    def focusLost(e: FocusEvent): Unit = {
      if (!e.isTemporary && e.getSource.isInstanceOf[ScrollableTextComponent])
        dontWatch()
    }
  }

  val FIND_ACTION = new FindAction
  val FIND_NEXT_ACTION = new FindNextAction

  val FIND_ACTION_CODE = new FindActionCode
  val FIND_NEXT_ACTION_CODE = new FindNextActionCode

  private var instance: FindDialog = null
  private var codeInstance: FindDialog = null

  def init(frame: Frame, codeFrame: Frame): Unit = {
    instance = new FindDialog(frame)
    codeInstance = new FindDialog(codeFrame)
  }

  def getInstance: FindDialog = {
    if (instance == null)
      throw new IllegalStateException("FindDialog was never initialized")

    instance
  }

  def getCodeInstance: FindDialog = {
    if (codeInstance == null)
      throw new IllegalStateException("FindDialog was never initialized")

    codeInstance
  }

  def watch(target: ScrollableTextComponent, code: Boolean = false): Unit = {
    if (code) {
      FIND_ACTION_CODE.setEnabled(true)
      getCodeInstance.target = Option(target)
      getCodeInstance.setReplaceEnabled(target.isEditable)
    }

    else {
      FIND_ACTION.setEnabled(true)
      getInstance.target = Option(target)
      getInstance.setReplaceEnabled(target.isEditable)
    }
  }

  def dontWatch(code: Boolean = false): Unit = {
    if (code) {
      getCodeInstance.setVisible(false)
      FIND_ACTION_CODE.setEnabled(false)
    }

    else {
      getInstance.setVisible(false)
      FIND_ACTION.setEnabled(false)
    }
  }

  override def syncTheme(): Unit = {
    if (instance != null)
      instance.syncTheme()

    if (codeInstance != null)
      codeInstance.syncTheme()

    FIND_ACTION.putValue(Action.SMALL_ICON, Utils.iconScaledWithColor("/images/find.png", 15, 15,
                                                                      InterfaceColors.toolbarImage()))
    FIND_ACTION_CODE.putValue(Action.SMALL_ICON, Utils.iconScaledWithColor("/images/find.png", 15, 15,
                                                                           InterfaceColors.toolbarImage()))
    FIND_NEXT_ACTION.putValue(Action.SMALL_ICON, Utils.iconScaledWithColor("/images/find.png", 15, 15,
                                                                           InterfaceColors.toolbarImage()))
    FIND_NEXT_ACTION_CODE.putValue(Action.SMALL_ICON, Utils.iconScaledWithColor("/images/find.png", 15, 15,
                                                                                InterfaceColors.toolbarImage()))
  }
}

class FindDialog(val owner: Frame) extends JDialog(owner, I18N.gui.get("dialog.find.title"), false) with ActionListener
                                                                                                    with ThemeSync {
  private var target: Option[ScrollableTextComponent] = None

  private val nextButton = new DialogButton(false, I18N.gui.get("dialog.find.next"), () => {
    if (!next(findBox.getText, ignoreCaseCheckBox.isSelected, wrapAroundCheckBox.isSelected)) {
      Toolkit.getDefaultToolkit.beep()
      notFoundLabel.setVisible(true)
    } else {
      notFoundLabel.setVisible(false)
    }
  })

  private val prevButton = new DialogButton(false, I18N.gui.get("dialog.find.previous"), () => {
    if (!prev(findBox.getText, ignoreCaseCheckBox.isSelected, wrapAroundCheckBox.isSelected)) {
      Toolkit.getDefaultToolkit.beep()
      notFoundLabel.setVisible(true)
    } else {
      notFoundLabel.setVisible(false)
    }
  })

  private val replaceButton = new DialogButton(false, I18N.gui.get("dialog.find.replace"), () => {
    replace(replaceBox.getText)
  })

  private val replaceAndFindButton = new DialogButton(false, I18N.gui.get("dialog.find.replaceAndFind"), () => {
    target.foreach { t =>
      if (t.getSelectedText != null && (
        if (ignoreCaseCheckBox.isSelected)
          t.getSelectedText.equalsIgnoreCase(findBox.getText)
        else
          t.getSelectedText.equals(findBox.getText)
      )) {
        replace(replaceBox.getText)
      }
    }

    if (!next(findBox.getText, ignoreCaseCheckBox.isSelected, wrapAroundCheckBox.isSelected)) {
      Toolkit.getDefaultToolkit.beep()
      notFoundLabel.setVisible(true);
    } else {
      notFoundLabel.setVisible(false);
    }
  })

  private val replaceAllButton = new DialogButton(false, I18N.gui.get("dialog.find.replaceAll"), () => {
    replaceAll(findBox.getText, ignoreCaseCheckBox.isSelected, replaceBox.getText)
  }: Unit)

  private val findBox = new TextField(25)
  private val replaceBox = new TextField(25)
  private val replaceLabel = new JLabel(I18N.gui.get("dialog.find.replaceWith"))
  private val notFoundLabel = new JLabel(I18N.gui.get("dialog.find.notFound"))

  new NonemptyTextFieldButtonEnabler(nextButton, List(findBox))
  new NonemptyTextFieldButtonEnabler(prevButton, List(findBox))

  private val replaceEnabler = new NonemptyTextFieldButtonEnabler(replaceButton, List(findBox))
  private val replaceAndFindEnabler = new NonemptyTextFieldButtonEnabler(replaceAndFindButton, List(findBox))
  private val replaceAllEnabler = new NonemptyTextFieldButtonEnabler(replaceAllButton, List(findBox))

  private val ignoreCaseCheckBox = new CheckBox(I18N.gui.get("dialog.find.ignoreCase")) {
    setSelected(true)
  }

  private val wrapAroundCheckBox = new CheckBox(I18N.gui.get("dialog.find.wrapAround")) {
    setSelected(true)
  }

  new NonemptyTextFieldActionEnabler(FindDialog.FIND_NEXT_ACTION, List(findBox))
  new NonemptyTextFieldActionEnabler(FindDialog.FIND_NEXT_ACTION_CODE, List(findBox))

  findBox.setEditable(true)
  replaceBox.setEditable(true)
  notFoundLabel.setVisible(false)

  getRootPane.setDefaultButton(nextButton)

  setResizable(false)
  setVisible(false)

  private val findPanel = new TextFieldBox(SwingConstants.LEFT)

  locally {
    findPanel.setBorder(new EmptyBorder(16, 8, 8, 8))

    findPanel.addField(I18N.gui.get("dialog.find.find"), findBox)
    findPanel.addField(replaceLabel, replaceBox)

    val optionsPanel = new JPanel with Transparent

    optionsPanel.setLayout(new BoxLayout(optionsPanel, BoxLayout.X_AXIS))
    optionsPanel.setBorder(new EmptyBorder(8, 8, 8, 8))

    optionsPanel.add(ignoreCaseCheckBox)
    optionsPanel.add(Box.createHorizontalStrut(12))
    optionsPanel.add(wrapAroundCheckBox)
    optionsPanel.add(Box.createHorizontalStrut(24))
    optionsPanel.add(notFoundLabel)

    val buttonPanel = new ButtonPanel(
      Seq(
        nextButton,
        prevButton,
        replaceButton,
        replaceAndFindButton,
        replaceAllButton
      ))

    buttonPanel.setBorder(new EmptyBorder(16, 8, 8, 8))

    getContentPane.setLayout(new BorderLayout)

    getContentPane.add(findPanel, BorderLayout.NORTH)
    getContentPane.add(optionsPanel, BorderLayout.CENTER)
    getContentPane.add(buttonPanel, BorderLayout.SOUTH)

    pack()
  }

  Utils.addEscKeyAction(this,
    new AbstractAction {
      def actionPerformed(e: ActionEvent): Unit = {
        setVisible(false)
      }
    })

  def actionPerformed(e: ActionEvent): Unit = {
    e.getSource match {
      case `findBox` =>
        if (!next(findBox.getText, ignoreCaseCheckBox.isSelected, wrapAroundCheckBox.isSelected)) {
          Toolkit.getDefaultToolkit.beep()
          notFoundLabel.setVisible(true)
        } else {
          notFoundLabel.setVisible(false)
        }

      case _ =>
        notFoundLabel.setVisible(false)
    }
  }

  // helper function for centering found or replaced text in the view (Isaac B 7/26/25)
  private def setScrollPosition(offset: Int): Unit = {
    target.foreach { t =>
      t.scrollPane.foreach { pane =>
        val pos = t.modelToView2D(offset)
        val xBar = pane.getHorizontalScrollBar
        val yBar = pane.getVerticalScrollBar

        xBar.setValue((pos.getX - xBar.getVisibleAmount / 2).toInt)
        yBar.setValue((pos.getY - yBar.getVisibleAmount / 2).toInt)
      }
    }
  }

  private def next(search: String, ignoreCase: Boolean, wrapAround: Boolean): Boolean = {
    target.map { t =>
      var searchMut = search
      var text = getTargetText

      if (ignoreCase) {
        // this might get slow with big programs. should be tested. -AZS
        searchMut = searchMut.toUpperCase(Locale.ENGLISH)
        text = text.toUpperCase(Locale.ENGLISH)
      }

      var matchIndex = text.indexOf(searchMut, t.getSelectionEnd)

      if (matchIndex == -1 && wrapAround)
        matchIndex = text.indexOf(searchMut)

      if (matchIndex > -1) {
        setScrollPosition(matchIndex)
        t.setSelectionStart(matchIndex)
        t.setSelectionEnd(matchIndex + searchMut.length)
        true
      } else {
        false
      }
    }.getOrElse(false)
  }

  private def prev(search: String, ignoreCase: Boolean, wrapAround: Boolean): Boolean = {
    target.map { t =>
      var searchMut = search
      var text = getTargetText

      if (ignoreCase) {
        // this might get slow with big programs. should be tested. -AZS
        searchMut = searchMut.toUpperCase(Locale.ENGLISH)
        text = text.toUpperCase(Locale.ENGLISH)
      }

      var matchIndex = text.lastIndexOf(searchMut, t.getSelectionStart - 1)

      if (matchIndex == -1 && wrapAround)
        matchIndex = text.lastIndexOf(searchMut)

      if (matchIndex > -1) {
        setScrollPosition(matchIndex)
        t.setSelectionStart(matchIndex)
        t.setSelectionEnd(matchIndex + searchMut.length)
        true
      } else {
        false
      }
    }.getOrElse(false)
  }

  private def replace(replacement: String): Unit = {
    target.foreach { t =>
      if (t.getSelectedText == null || t.getSelectedText.isEmpty) {
        Toolkit.getDefaultToolkit.beep()
      } else {
        try {
          setScrollPosition(t.getSelectionStart)
          t.getDocument.remove(t.getSelectionStart, t.getSelectionEnd - t.getSelectionStart)
          t.getDocument.insertString(t.getCaretPosition, replacement, null)
        } catch {
          case ex: BadLocationException =>
            Toolkit.getDefaultToolkit.beep()
        }
      }
    }
  }

  private def replaceAll(search: String, ignoreCase: Boolean, replacement: String): Int = {
    target.map { t =>
      t.setSelectionStart(0)
      t.setSelectionEnd(0)

      if (next(search, ignoreCase, false)) {
        var i = 1

        while {
          replace(replacement)

          i += 1

          if (i > 50000)
            throw new IllegalStateException("Replace All replaced too many items.")

          next(search, ignoreCase, false) // never wrap around on replace all
        } do ()

        i
      } else 0
    }.getOrElse(0)
  }

  private def getTargetText: String = {
    target match {
      case Some(editor: JEditorPane) =>
        // we need to get the text this way to avoid returning the HTML
        // tags which screw-up the search - jrn 7/22/05
        try {
          editor.getText(0, editor.getDocument.getLength)
        } catch {
          case ex: BadLocationException =>
            throw new IllegalStateException(ex)
        }

      case Some(t) =>
        t.getText

      case _ => ""
    }
  }

  private def setReplaceEnabled(enabled: Boolean): Unit = {
    replaceEnabler.setEnabled(enabled)
    replaceAndFindEnabler.setEnabled(enabled)
    replaceAllEnabler.setEnabled(enabled)
    replaceBox.setEnabled(enabled)
    replaceLabel.setEnabled(enabled)
  }

  override def syncTheme(): Unit = {
    getContentPane.setBackground(InterfaceColors.dialogBackground())

    nextButton.syncTheme()
    prevButton.syncTheme()
    replaceButton.syncTheme()
    replaceAndFindButton.syncTheme()
    replaceAllButton.syncTheme()

    ignoreCaseCheckBox.setForeground(InterfaceColors.dialogText())
    wrapAroundCheckBox.setForeground(InterfaceColors.dialogText())

    findBox.syncTheme()
    replaceBox.syncTheme()

    replaceLabel.setForeground(InterfaceColors.dialogText())
    notFoundLabel.setForeground(InterfaceColors.dialogText())

    findPanel.syncTheme()
  }
}
